File: assignments/assignment-2-solution/src-csharp/TeaShop/Domain/Query/Sorts/SortByPriceDecorator.cs
using Assignment2Solution.Domain.Inventory;

namespace Assignment2Solution.Domain.Query.Sorts;

/// <summary>
///     A decorator that sorts inventory items by price.
/// </summary>
public sealed class SortByPriceDecorator : InventoryQueryDecoratorBase
{
    private readonly SortDirection _direction;

    /// <summary>
    ///     Initializes a new instance of the <see cref="SortByPriceDecorator" /> class.
    /// </summary>
    /// <param name="inner">The inner query to decorate.</param>
    /// <param name="direction">The sort direction.</param>
    public SortByPriceDecorator(IInventoryQuery inner, SortDirection direction = SortDirection.Ascending)
        : base(inner)
    {
        _direction = direction;
    }

    /// <inheritdoc />
    protected override string? AppliedDescription
        => $"Sort: Price ({_direction.ToString().ToLowerInvariant()})";

    /// <inheritdoc />
    public override IReadOnlyList<InventoryItem> Execute()
    {
        var items = Inner.Execute();

        var sorted = _direction == SortDirection.Ascending
            ? items.OrderBy(i => i.Price)
            : items.OrderByDescending(i => i.Price);

        return sorted.ToList().AsReadOnly();
    }
}

File: assignments/assignment-2-solution/src-java/src/main/java/assignment2solution/domain/inventoryquery/sorts/SortByPriceDecorator.java
package assignment2solution.domain.inventoryquery.sorts;

import assignment2solution.domain.inventory.InventoryItem;
import assignment2solution.domain.inventoryquery.IInventoryQuery;
import assignment2solution.domain.inventoryquery.InventoryQueryDecoratorBase;

import java.util.Comparator;
import java.util.List;
import java.util.Locale;
import java.util.stream.Collectors;

/**
 * A decorator that sorts inventory items by price.
 */
public final class SortByPriceDecorator extends InventoryQueryDecoratorBase {
    private final SortDirection direction;

    /**
     * Initializes a new instance of the {@link SortByPriceDecorator} class.
     */
    public SortByPriceDecorator(IInventoryQuery inner, SortDirection direction) {
        super(inner);
        this.direction = direction == null ? SortDirection.ASCENDING : direction;
    }

    public SortByPriceDecorator(IInventoryQuery inner) {
        this(inner, SortDirection.ASCENDING);
    }

    @Override
    protected String getAppliedDescription() {
        return "Sort: Price (" + direction.name().toLowerCase(Locale.ROOT) + ")";
    }

    @Override
    protected List<InventoryItem> decorate(List<InventoryItem> items) {
        var comparator = Comparator.comparing(InventoryItem::getPrice);
        if (direction == SortDirection.DESCENDING) {
            comparator = comparator.reversed();
        }
        return items.stream().sorted(comparator).collect(Collectors.toUnmodifiableList());
    }
}
